DSP-based tempo detection

saintmatthieu opened the pull request:
Resolves: #5640

DSP method for the detection of musical meter, and its integration.

<!-- Use "x" to fill the checkboxes below like [x] -->

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [ ] Each commit compiles and runs on my machine without known undesirable changes of behavior

# QA:

## New items

- [x] Importing very short file do not crash
- [x] Importing files longer than 1mn does not trigger the detection

## From title-parsing ticket #5570

This is a selection of the QA items found #5570 that might have been affected by this PR.

### UX flow and tempo-detection side effects
#### Importing with "Ask me each time"

The user has their pref set to "Ask me each time",
- [x] importing on empty project launches user prompt
- [x] importing on a non-empty project, if the view is in minutes and seconds, nothing happens
- [x] importing on non-empty project, if the view is in beats and measures, clip is auto-stretched, and Ctrl-Z undoes import

#### Interacting with modal

The project is empty, the user hasn't already recorded their preference by checking "Don't ask again", and imports a file with BPM in its title. The prompt appears.

-  [x] If a user says "No", auto-stretches if in B&M mode, else nothing happens. Ctrl-Z undoes import.
-  [x] If a user says "Yes", the view switches to Beats-and-measures if not already there, and the project tempo becomes that of the clip. Ctrl-Z undoes only project configuration, without stretching the clip.

#### Persistency

- [x] The user recorded their preference by checking "Don't ask again". They now drop an audio file with BPM information in its title. Audacity behaves as described in this table:

| Case | Pref         | View is Beats and Measures | Project is empty | Result                                     |
| ---- | ------------ | -------------------------- | ---------------- | ------------------------------------------ |
| A    | "Do nothing" | No                         | Yes              | Nothing                                    |
| B    | "Do nothing" | No                         | No               | Nothing                                    |
| C    | "Do nothing" | Yes                        | Yes              | auto-stretch                               |
| D    | "Do nothing" | Yes                        | No               | auto-stretch                               |
| E    | "Switch ..." | No                         | Yes              | Switch to B&M and project takes clip tempo |
| F    | "Switch ..." | No                         | No               | Nothing                                    |
| G    | "Switch ..." | Yes                        | Yes              | project takes clip tempo                   |
| H    | "Switch ..." | Yes                        | No               | auto-stretch                               |

#### Miscellaneous
- [x] Answering "Yes" to the prompt adds a history item, and can be undone from the `Edit` menu or with ctrl-Z. Undoing means reverting to the previous tempo and view.
- [x] The amount of stretching is minimized through power-of-two speed changes. E.g., if a clip with title BPM 50 is dropped in a project with tempo 90, the clip is auto-stretched to 45 BPM, i.e., its speed is set to 90% rather than 180%.
- [x] For loops lasting _N_ quarter notes + _e_, with _e_ less than a 32nd, _e_ is trimmed off from the right (to ease horizontally duplicating loops)
- [x] Importing more than one file at once behaves like it did before - no stretching, no project re-configuration.

saintmatthieu commented:
@rbdannenberg this might be of interest to you? In any case, this whole initiative would benefit from your input.

saintmatthieu commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+// Look at the amplitude across all scores. If our ODF is that of a musical
+// rhythm, then the contrast between good and bad quantizations should make a
+// big difference.
+double GetExperimentContrastMeasure(
```
I just removed the contrast measure in a follow-up commit, improving performance of the classifier and simplifying the procedure. Please ignore this part while reviewing.

Paul-Licameli commented on tests/Mp3FileReader.cpp:
```diff
@@ -0,0 +1,71 @@
+
+  Audacity: A Digital Audio Editor
+
+  WavFileIO.cpp
```
Fix the comment

Paul-Licameli commented on tests/Mp3FileReader.h:
```diff
@@ -0,0 +1,28 @@
+
+  Audacity: A Digital Audio Editor
+
+  WavFileIO.h
```
fix the comment

Paul-Licameli commented on libraries/lib-math/LinearFit.h:
```diff
@@ -0,0 +1,41 @@
+#include <cassert>
```
Put a banner on the file and `#include` guard

Paul-Licameli commented on libraries/lib-math/LinearFit.h:
```diff
@@ -0,0 +1,41 @@
+
+template <typename T, typename U>
+std::pair<double, double> LinearFit(
+   const std::vector<T>& x, const std::vector<U>& y, std::vector<double> w = {})
```
I wish we had `std::span` from C++20.  I wish we could use `std::vector` less often to pass contiguous ranges of values.  Because `std::vector` specifies storage management details that are extraneous to the algorithm.

Without `span` you would need at least one extra `size_t` argument so that you could pass simple pointers.

Don't fix this but remember it and think about it.

Paul-Licameli commented on libraries/lib-math/LinearFit.h:
```diff
@@ -0,0 +1,41 @@
+std::pair<double, double> LinearFit(
+   const std::vector<T>& x, const std::vector<U>& y, std::vector<double> w = {})
+{
+   assert(x.size() == y.size() && (w.empty() || w.size() == y.size()));
```
The assertion should be commented as a precondition

Paul-Licameli commented on libraries/lib-math/LinearFit.h:
```diff
@@ -0,0 +1,41 @@
+#include <vector>
+
+template <typename T, typename U>
+std::pair<double, double> LinearFit(
```
You use template parameters T and U, not requiring them to be the same.

You use `double` in the return type.

Maybe you might make this more general?  (Could there be some future need to use it with `long double` or `std::complex<float>?)

Maybe

`template<typename ResultType = std::common_type_t<double, T, U>, typename T, typename U>`

Maybe also the common cases where T and U are both double or both float could also declare `extern MATH_API` explicit template instantiations here, and a companion .cpp file could generate the definitions.

saintmatthieu replied:
I do think the case where `x` is of integer type is quite common. That's the use case I had, rounding onsets to tatum (or "tick") indices.
For the rest, it's a bit complicated to find a solution that satisfies all cases without compiler warnings. I just generalized the weight type for usage convenience, but the return type is just double. Users who need float coefficients shouldn't have trouble downcasting.

Paul-Licameli commented on libraries/lib-math/LinearFit.h:
```diff
@@ -0,0 +1,41 @@
+      x.begin(), x.end(), y.begin(), 0.0, std::plus<>(),
+      [&](double xi, double yi) {
+         return w[n++] * (xi - w_mean_x) * (yi - w_mean_y);
+      });
```
See here for a formula that suggests a one-pass accumulation of the answer which might or might not make a difference in performance https://en.wikipedia.org/wiki/Covariance#Mathematics

But see also https://en.wikipedia.org/wiki/Catastrophic_cancellation

Paul-Licameli commented on libraries/lib-math/LinearFit.h:
```diff
@@ -0,0 +1,41 @@
+      x.begin(), x.end(), x.begin(), 0.0, std::plus<>(),
+      [&](double xi, double) {
+         return w[n++] * (xi - w_mean_x) * (xi - w_mean_x);
+      });
```
Here too, this is just variance (times N)

Paul-Licameli commented on src/CMakeLists.txt:
```diff
@@ -1317,6 +1319,7 @@ set( AUDACITY_LIBRARIES
    lib-note-track-interface
    lib-viewport-interface
    lib-music-information-retrieval-interface
+   lib-wave-track-interface
```
I think this line is unnecessary to build

Paul-Licameli commented on src/ClipMirAudioReader.cpp:
```diff
@@ -0,0 +1,40 @@
+#include "ClipMirAudioReader.h"
```
Banners for this file and the header

Paul-Licameli commented on src/ClipMirAudioReader.cpp:
```diff
@@ -0,0 +1,40 @@
+void ClipMirAudioReader::ReadFloats(
+   float* buffer, long long where, size_t numFrames) const
+{
+   AddChannel(0, buffer, where, numFrames);
```
Previous contents of `buffer` matter -- this is addition, not assignment.  A header comment should say that.

Paul-Licameli replied:
I took another look this morning, and I'm still not convinced, that the loop using `GetNextFrame` in `MirDsp.cpp` is correct.  I still think later passes have a buffer contaminated with previous data.

But all of that could be fixed down at this level, if only the first `AddChannel` instead assigned the buffer, not accumulating into it.  Only the second `AddChannel` should remain.  And then no special header comments would be needed.

saintmatthieu replied:
Yes indeed, this wasn't intended behaviour. Addressing.

Paul-Licameli replied:
I see it is now corrected, though there are two passes over the buffer (to zero it, then to add to it).  Maybe that marginal inefficiency is unimportant.

Paul-Licameli commented on libraries/lib-music-information-retrieval/StftFrameProvider.cpp:
```diff
@@ -0,0 +1,91 @@
+{
+   if (mNumFramesProvided >= mNumFrames)
+      return false;
+   frame.resize(mFftSize);
```
I figured out that the only uses of `GetFloats` are in this function.  Is there a guarantee that `frame.data()` used as the buffer will be zero-initialized first?  Only if the given `frame` is empty before this `resize`.  Precede this with `resize(0)` and then there is no more doubt.

saintmatthieu replied:
https://github.com/audacity/audacity/pull/5791#discussion_r1437562009

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+   std::fill(prevPowSpec.begin(), prevPowSpec.end(), 0.f);
+
+   auto frameCounter = 0;
+   while (frameProvider.GetNextFrame(buffer))
```
Here is the use of `GetNextFrame` but I see no zeroing out of the buffer between passes.

Is there in fact a bug here?

saintmatthieu replied:
https://github.com/audacity/audacity/pull/5791#discussion_r1437562009

Paul-Licameli commented on libraries/lib-music-information-retrieval/StftFrameProvider.cpp:
```diff
@@ -0,0 +1,91 @@
+   // It's not impossible that some user drops a file so short that `mFftSize >
+   // mNumSamples`. In that case we won't be returning a meaningful
+   // STFT, but that's a use case we're not interested in. We just need to make
+   // sure we don't crash.
```
To treat the short file boundary case really fully, should there be a while loop below?  With a careful test that the loop makes progress each time?  Or should you be sure to zero-fill any uninitialized part remaining after the second read?  (Which would be a consequence of first resizing the buffer to zero above)

saintmatthieu replied:
Yes, the perfect circular read-out would need a while loop. It probably makes implementation easier not to care for this use case, which would be too short to be a musical loop anyway.
I think initialization to 0 would only improve situations where the underlying file is less than twice the FFT size, but it's cheap and easy, so will do.

Paul-Licameli commented on libraries/lib-music-information-retrieval/StftFrameProvider.h:
```diff
@@ -0,0 +1,28 @@
+{
+class MirAudioReader;
+
+class StftFrameProvider
```
Clarify with comments that this class does not calculate STFT, but is only a source of time-domain frames

Paul-Licameli commented on libraries/lib-stretching-sequence/AudioSegmentSampleView.cpp:
```diff
@@ -39,25 +40,39 @@ void AudioSegmentSampleView::Copy(float* buffer, size_t bufferSize) const
+      const auto dst = buffer + written;
+      std::transform(
+         src, src + toWriteFromBlock, dst, dst,
+         [](float a, float b) { return a + b; });
```
You might just write `std::plus{}`

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirTypes.h:
```diff
@@ -0,0 +1,56 @@
+
+// Maps a combination of time-signature and number of bars (encoded as
+// string) to a score.
+struct OnsetQuantization
```
Not clear in context here, what string you mean.

Use `//!` instead of `//` and it becomes a Doxygen brief comment.

saintmatthieu replied:
That's an outdated comment - thanks.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirUtils.h:
```diff
@@ -0,0 +1,52 @@
+
+constexpr auto MapToPositiveHalfIndex(int index, int fullSize)
+{
+   assert(fullSize % 2 == 0);
```
Make a `@pre` comment

Paul-Licameli replied:
Is there also an intended assumption that `index < fullSize` ?

Paul-Licameli replied:
Are you also assuming `fullSize > 0` ?  (Loops below won't terminate if it is 0)

saintmatthieu replied:
Making it unsigned doesn't protect against invalid input, since `fullSize > 0` is a pre-condition, so I don't see that much of a benefit.
But yes, I should handle invalid argument values.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirUtils.h:
```diff
@@ -0,0 +1,52 @@
+   while (index < 0)
+      index += fullSize;
+   while (index >= fullSize)
+      index -= fullSize;
```
I might write
```
if (index >= 0)
   index = index % fullSize;
else
   index = fullSize - (-index % fullSize);
```

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirUtils.h:
```diff
@@ -0,0 +1,52 @@
+
+constexpr auto IsPowOfTwo(int x)
+{
+   return (x & (x - 1)) == 0;
```
Incorrect if x is 0 or negative; again consider an unsigned argument

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirUtils.cpp:
```diff
@@ -0,0 +1,80 @@
+
+// Function to generate numbers whose prime factorization contains only twos or
+// threes
+std::vector<int> GetPowersOf2And3(int lower, int upper)
```
The function name doesn't suggest exactly what it does.  6 is not a power of 2 or of 3.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirUtils.cpp:
```diff
@@ -0,0 +1,80 @@
+   result.erase(
+      std::remove_if(
+         result.begin(), result.end(),
+         [](int n) { return CountThreesInPrimeDecomposition(n) > 2; }),
```
Or simply `n % 9 == 0`

Paul-Licameli replied:
So really `GetPowersOf2And3` could generate just the powers of 2 and the triples of those

saintmatthieu replied:
`n % 27 == 0` - we hold musical rhythms with two triple levels for realistic (e.g. a swung 6/8). But thanks for the tip.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirUtils.cpp:
```diff
@@ -0,0 +1,80 @@
+   return result;
+}
+
+std::vector<int> GetPeakIndices(const std::vector<float>& odf)
```
I don't know what `odf` abbreiviates

Paul-Licameli replied:
Onset detection function

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirUtils.cpp:
```diff
@@ -0,0 +1,80 @@
+   std::vector<int> peakIndices;
+   for (auto j = 1; j < odf.size() - 1; ++j)
+   {
+      const auto i = j == 0 ? odf.size() - 1 : j - 1;
```
`j` is never 0

saintmatthieu replied:
Right - and this was a bug, missing the presence of a peak at index 0. Thank you, addressing this has improved the performance of the classifier. See c0b0d85123b5403d4c8355ffdfc92374c4b46da0

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirAudioReader.h:
```diff
@@ -0,0 +1,16 @@
+   virtual long long GetNumSamples() const = 0;
+   virtual void
+   ReadFloats(float* buffer, long long where, size_t numFrames) const = 0;
+   virtual ~MirAudioReader() = default;
```
I would make a small companion `.cpp` file, just to define the destructor non-inline, so that the linker doesn't need to de-duplicate the virtual function table but will allocate it just in that .cpp.

Not very important for a header file that is not widely used, but I make a habit of this whenever I make a new abstract base class.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MusicInformationRetrieval.h:
```diff
@@ -10,12 +10,38 @@
 namespace MIR
 {
+class MirAudioReader;
+struct BeatInfo;
```
This type is nowhere used

saintmatthieu replied:
An obsolete remnant.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MusicInformationRetrieval.h:
```diff
@@ -10,12 +10,38 @@
+   const double allowedFalsePositiveRate;
+
+   /*!
+    * Classifier score threshold above which the analyzed audio file can be
```
incomplete comment?

Paul-Licameli commented on libraries/lib-music-information-retrieval/MusicInformationRetrieval.h:
```diff
@@ -10,12 +10,38 @@
+   /*!
+    * Classifier score threshold above which the analyzed audio file can be
+    */
+   const double threshold;
```
If I correctly understand the statistical methods, then whether the allowed false positive rate is consistent with the threshold, depends on probability distributions of ... whatever `threshold` measures, in two different populations of sound files, those that have rhythmic content and those that do not.

So then are the four magic numbers below derived experimentally from a certain corpus of example files that you take as representative?

saintmatthieu replied:
Two distributions of the score returned by the algorithm, yes, one for files whose ground truths say they are not music loops, and the other for the others. If the distributions do not overlap, then you have a perfect classifier and the threshold can be set anywhere between the two. But ours do, and we set the threshold such that it splits the non-loop population in 99% / 1% for the strict case, and 90% / 10% for the lenient case, the upper part of the split being the false-positive area.
Representativeness is indeed a topic. Ideally I could take a representative sample set from what users are importing, but I don't have this. So I need a data set I'm happy with, and believe in it. The one use in this PR (next commit) is worth what it's worth. A follow-up ticket, #5795, may provide something more representative (although the goal of that ticket is first and foremost to make the dataset available to others for reproducibility).

Paul-Licameli commented on libraries/lib-music-information-retrieval/MusicInformationRetrieval.cpp:
```diff
@@ -9,35 +9,75 @@
+            meter.has_value())
+   {
+      bpm = meter->bpm;
+      timeSignature = meter->timeSignature;
```
`timeSignature` is not assigned on all paths.  Better to make it consistent that way.

Or maybe better, why not return `MusicalMeter` by value and eliminate arguments?

saintmatthieu replied:
Absolutely.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MusicInformationRetrieval.cpp:
```diff
@@ -9,35 +9,75 @@
+   FillMetadata(
+      filename, audio, tolerance, std::move(progressCallback),
+      const_cast<std::optional<double>&>(mBpm),
+      const_cast<std::optional<TimeSignature>&>(mTimeSignature));
```
Maybe one `MusicalMeter` member instead, directly initialized

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirUtils.cpp:
```diff
@@ -0,0 +1,80 @@
+{
+namespace
+{
+constexpr auto bpmStdDev = 29.7953;
```
Should this standard deviation be defined in the same place as the mean, `bpmExpectedValue`?

saintmatthieu replied:
These will be subject to change in follow-up improvement PRs, I'm reluctant to invest time centralizing them, as little work as that may be.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.h:
```diff
@@ -0,0 +1,41 @@
+struct QuantizationFitDebugOutput;
+
+/*!
+ * @brief Get the normalized, circular auto-correlation of a signal for a signal
```
Redundant words

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+   constexpr auto smoothingWindowDuration = 0.2;
+   // An odd number.
+   const int M = std::round(smoothingWindowDuration * sampleRate / 4) * 2 + 1;
+   const auto window = GetNormalizedHann(2 * M + 1);
```
This use of the normalized Hann window is only as a moving-average kernel -- spectral properties are not important.  Am I right?

saintmatthieu replied:
I suppose it is in some way. The average of the signal is like its DFT at 0Hz. The window function reduces the amount of leakage of non-zero components in the 0Hz measurement.

saintmatthieu replied:
... but it's not meant for improved frequency-domain analysis, if that's what you meant.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+         else if (k >= x.size())
+            k -= x.size();
+         y += x[k] * window[i + M];
+      }
```
Maybe that loop can just be `std::accumulate` on one or two ranges

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+         y += x[k] * window[i + M];
+      }
+      ++n;
+      constexpr auto smoothingThreshold = 1.5;
```
I don't understand this multiplier.

saintmatthieu replied:
It's a poorly documented yet important multiplier.
The moving average of the raw ODF will be subtracted from it to yield the final ODF, negative results being set to 0. This is to remove noise of small ODF peaks before the method's quantization step.
Changing this value can impact dramatically the performance of the classifier.
I'm still not completely happy with the ODF. In some cases it can be very noisy in comparison with our perception of where onsets are. I'm exploring better solutions, and if I do find one, this smoothing may become obsolete.
Still, adding a comment.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+std::vector<float> GetNormalizedCircularAutocorr(std::vector<float> x)
+{
+   const auto N = x.size();
+   assert(IsPowOfTwo(N));
```
Make a `@pre`

saintmatthieu replied:
It's in the header.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+   InverseRealFFT(N, x.data(), nullptr, x.data());
+   // For efficiency, only keep the positive half of this symmetric signal.
+   x.erase(x.begin() + N / 2 + 1, x.end());
+   const auto normalizer = 1 / x[0];
```
Can this divide by 0?

saintmatthieu replied:
Not unless the input is only zeroes. Adding a check.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+}
+
+std::vector<float>
+GetMovingAverage(const std::vector<float>& x, double sampleRate)
```
The argument is called `sampleRate` but at the point of usage it is passed a frame rate, which is sample rate divided by a hop size.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+      std::transform(
+         powSpec.begin(), powSpec.end(), powSpec.begin(),
+         // Using `logf` on Linux fails. #ifdef it ?
+         [gamma](float x) { return std::log(1 + gamma * std::sqrt(x)); });
```
I don't pretend to know the theory here.  Maybe school me later.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+   const auto numFrames = frameProvider.GetNumFrames();
+   const auto frameSize = frameProvider.GetFftSize();
+   std::vector<float> buffer(frameSize);
+   std::vector<float> odf;
```
Maybe reserve `frameProvider.GetNumFrames()` ?

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+         debugOutput->postProcessedStft.push_back(powSpec);
+
+      if (progressCallback)
+         progressCallback(1. * ++frameCounter / numFrames);
```
If performance is concern enough for a progress indicator, perhaps we should seek optimizations.  Maybe the old functions in FFT.cpp could be improved.  Such as by avoiding the un-permuting and re-permuting of the spectrum.

saintmatthieu replied:
#5794 is planned for optimization. Downsampling to 16kHz is already something that could augment speed drastically.
But I'm keeping that ticket for later. Like I said, I'm not completely happy with the ODF, and reworking it would put any optimization at risk.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.h:
```diff
@@ -0,0 +1,82 @@
+ * method.
+ *
+ * @param audio The audio file to analyze.
+ * @param view The view in which the audio file is displayed.
```
No such parameter

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+   constexpr auto maxBarDuration = 4.;
+   const int minNumBars =
+      std::max(std::round(audioFileDuration / maxBarDuration), 1.);
+   const int maxNumBars = std::round(audioFileDuration / minBarDuration);
```
Is minNumBars <= maxNumBars always?

saintmatthieu replied:
Will `IotaRange` complain if it isn't?

Paul-Licameli replied:
It may loop infinitely because the termination test is == not >=

Paul-Licameli replied:
So write `const int maxNumBars = std::min(minNumBars, ...);`

saintmatthieu replied:
`minBarDuration < maxBarDuration`, and they divide the same value, so `minNumBars <= maxNumBars`.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+                  const auto tatumRate = numTatums / audioFileDuration * 60;
+                  if (
+                     IsRound(tatumsPerBar) && minTatumsPerMinute < tatumRate &&
+                     tatumRate < maxTatumsPerMinute)
```
<= not < ? (twice)

saintmatthieu replied:
Given that `tatumRate` is not an integer, that won't make much difference. `minTatumsPerMinute` and the other are ballpark values anyway.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+   const int minNumBars =
+      std::max(std::round(audioFileDuration / maxBarDuration), 1.);
+   const int maxNumBars = std::round(audioFileDuration / minBarDuration);
+   std::vector<int> numBars(maxNumBars - minNumBars + 1);
```
I wish we had C++20 ranges.  Without them, you could use standard algorithms without needing these temporary vectors just to hold integers.

Making these vectors seems wasteful and only obscures the two outer loops that might just as easily be written with an integer loop variable.

Paul-Licameli commented:
Please note, I have pushed a few suggested commits onto the branch to simplify some algorithms.  I cherry-picked IteratorX.h from the wave-clip-refactoring branch, which you reviewed before.  I extended it with more utilities.  You can avoid temporary vectors of integers just to use iota.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+// more accurate the performer has to be for this distance to stay low. In
+// general, this is desired. We could nevertheless consider introducing a slight
+// tolerance which increases with the tatum rate, as very fast rhythmic patterns
+// recorded without post-editing still tend to achieve lower scores.
```
This comment makes me think of application of a varying stretch to correct irregularities of rhythm... have you thought of that?

saintmatthieu replied:
This could be an application of the result of this analysis indeed. We could offer the user the possibility to automatically warp the onsets to the nearest tatum. Generalizing our integration of the time-stretching from StaffPad may not be that tricky, but there would be real difficulties related to waveform painting. Or maybe the painting would be so subtle we wouldn't even need to address it ... then pencil and brush would also have to be disabled for time-warped clips, but UX-wise that wouldn't be too bad either.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+{
+   std::vector<double> peakDistances(peakIndices.size());
+   const auto peakSum = std::accumulate(
+      peakValues.begin(), peakValues.end(), 0., std::plus<double> {});
```
`std::accumulate` defaults the operation to addition, so you can shorten that

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+         const auto distance = 2 *
+                               (shiftedIndex - closestDiv * odfSamplesPerDiv) /
+                               odfSamplesPerDiv;
+         return distance < 0 ? -distance : distance;
```
`std::abs`

saintmatthieu replied:
Of course.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+         const auto shiftedIndex = peakIndex - lag;
+         const auto closestDiv = std::round(shiftedIndex / odfSamplesPerDiv);
+         // Normalized distance between 0 and 1:
+         const auto distance = 2 *
```
Why `2 *`?

saintmatthieu replied:
Such that the error nicely spans `[0, 1)`. Will comment.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+         [](double a, double b) { return a + b; },
+         [&](double distance, float peakValue) {
+            return distance * peakValue;
+         }) /
```
You can default the last two arguments

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+            const auto distance = GetQuantizationDistance(
+               peakIndices, peakValues, odf.size(), numTatums, lag);
+            return std::make_pair(
+               numTatums, OnsetQuantization { distance, lag, numTatums });
```
It looks like you just want an `unordered_set` hashed on the third member

saintmatthieu replied:
I don't understand.

Paul-Licameli replied:
`numTatums` is the key of the map but also duplicated as part of the value.

It's not important.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+         return a.second.error < b.second.error;
+      });
+
+   // Read the normalized distance, a value between 0 and 1, as a probability.
```
Is the normalized distance really a "probability" (vertical axis if you plot the distribution) or is it the "random variable" (horizontal axis)?

saintmatthieu replied:
You're right, it's not a probability, `1 - distance` it's just a score. The probability of the file being a loop given that score is what we use in the benchmarking to determine the thresholds.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+   // Read the normalized distance, a value between 0 and 1, as a probability.
+   const auto error = bestFitIt->second.error;
+   const auto contrast = GetExperimentContrastMeasure(quantizations, error);
+   // The higher the contrast, the more likely the recording is a loop.
```
Explain "contrast" -- the same thing as the "score"?

saintmatthieu replied:
Outdated comment, thanks. (See https://github.com/audacity/audacity/pull/5791#discussion_r1431719151 if you wish)

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+   const auto mostLikelyNumTatums = bestFitIt->first;
+   const auto lag = bestFitIt->second.lag;
+
+   return { error, lag, mostLikelyNumTatums, score };
```
I would write inner braces around the first three members

saintmatthieu replied:
In fact `score` and `error` are rather redundant. I simplified.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,476 @@
+      return quantizations;
+   }();
+
+   const auto bestFitIt = std::min_element(
```
Is it correct to choose the hypothesis that minimizes error as the most likely correct hypothesis -- or, is there need to do something fancier with a Bayesian formula?  Which would make some assumptions about the unconditional probabilities of the hypotheses (how often will a loop have 6/8? how often 4/4?  etc.) and those data again might need to be derived from some corpus of examples we take as representative.

Paul-Licameli replied:
e.g. find a large enough population of examples where time signature is deduced from explicit metadata, not calculation?

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
What is the source of all the numbers in these three maps?

Paul-Licameli replied:
The data in the third map -- however derived -- might be used in that suggested Bayesian formula.

saintmatthieu replied:
You're right, I am using probabilistic measures that may be more clearly be brought together with a clear formulation of the Bayes theorem. I just created a ticket for that: https://github.com/audacity/audacity/issues/5826

Paul-Licameli replied:
I still don't know the source of the numbers.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
That has units of samples / beat, so I might call it `beatLength`

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
You are testing increasing periods, not frequencies -- maybe that should really be called "sub-harmonic"

saintmatthieu replied:
It's more the period index that I'm increasing. Renaming.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
I think you could sequence this test before all the rest of the calculation of autocorrelation and scores, and skip all that if it is true

saintmatthieu replied:
Nice! That may be a non-negligible speed improvement. I may remove this `fourFourIt` in a follow-up algorithmic improvement, but until then ...

saintmatthieu commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,159 @@
+      const auto y =
+         std::accumulate(m.begin(), m.end(), 0., [&](double y, int i) {
+            auto k = n + i;
+            if (k < 0)
```
This won't help for very small x.size().

Paul-Licameli replied:
I don't understand the context of the remark

saintmatthieu replied:
Sorry, was just a not for myself. If `x.size() == 1`, this crashes. Have to fix this.

Paul-Licameli commented on libraries/lib-music-information-retrieval/StftFrameProvider.cpp:
```diff
@@ -57,7 +57,7 @@ bool StftFrameProvider::GetNextFrame(std::vector<float>& frame)
    if (mNumFramesProvided >= mNumFrames)
       return false;
-   frame.resize(mFftSize);
+   frame.resize(mFftSize, 0.f);
```
Maybe that line is unnecessary, the previous fix being enough?

Paul-Licameli commented on libraries/lib-music-information-retrieval/StftFrameProvider.cpp:
```diff
@@ -35,8 +35,10 @@ double GetHopSize(int sampleRate, long long numSamples)
    const auto idealHopSize = 0.01 * sampleRate;
-   const auto numFrames =
-      1 << (int)std::round(std::log2(numSamples / idealHopSize));
+   const int exponent = std::round(std::log2(numSamples / idealHopSize));
```
Are you trying to handle corner cases?  You deal with zero for `numSamples`.

log2 returns -infinity for a zero argument, NaN for a negative, and idealHopSize might be zero.  I'm just trying to think of abusive inputs... the numbers come from virtual functions of `audio` below which have no postconditions.

Paul-Licameli commented:
I figured out the reason for the crash in the unit test in macOS release builds, and I presume the Linux builds were failing for the same reason.

Paul-Licameli commented on libraries/lib-music-information-retrieval/StftFrameProvider.h:
```diff
@@ -0,0 +1,28 @@
+class StftFrameProvider
+{
+public:
+   StftFrameProvider(const MirAudioReader& source);
```
Considering that crash I fixed in the unit test -- you might comment that the lifetime of the source encloses that of this?

Paul-Licameli commented on libraries/lib-music-information-retrieval/MusicInformationRetrieval.cpp:
```diff
@@ -71,7 +113,8 @@ ProjectSyncInfo MusicInformation::GetProjectSyncInfo(
    if (0 < delta && delta / 8)
```
Did you mean `0 < delta && delta < 1. / 8`

Paul-Licameli commented on src/ProjectFileManager.cpp:
```diff
@@ -1397,16 +1421,24 @@ void ReactOnMusicFileImport(
          else if (isBeatsAndMeasures)
             clip->StretchBy(syncInfo.stretchMinimizingPowOfTwo);
          if (!ans.readFromPreference && ans.yes)
             UndoManager::Get(*pProj).PushState(
```
This may be a problem:  The function may take paths that do change the tracks, but do not push an undo state.  Therefore those changes of the tracks will be lost if you Undo the Import command that was previously pushed, then redo again.  Or the changes will be consolidated as part of the next Undo item that is pushed, which may be for an unrelated editing operation.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+      std::swap(prevPowSpec, powSpec);
+
+      if (debugOutput)
+         debugOutput->postProcessedStft.push_back(powSpec);
```
It's only for debug purposes.  Do you want to push `powSpec` before or after the swap?

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,138 @@
+
+   // Close the loop.
+   odf.push_back(GetNoveltyMeasure(prevPowSpec, firstPowSpec));
+   assert(IsPowOfTwo(odf.size()));
```
Do you want to push one more power spectrum for debug output?

saintmatthieu replied:
Actually not: we are pushing another ODF frame because it's based on the adjacent differences. The correct number of power spectra already has been pushed.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,159 @@
+            if (k < 0)
+               k += x.size();
+            else if (k >= x.size())
+               k -= x.size();
```
`const auto size = x.size()` (lift that out of the loop)
`k = (k + n + i + size()) % size()`

I don't know whether that might make this hot inner loop run faster -- measure

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.h:
```diff
@@ -0,0 +1,77 @@
+BPM likelihood are combined together, and the BPM with largest score is
+returned.
+
+This approach is in some aspects alike existing tempo detection methods (e.g.
```
like not alike

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
`const auto odfSr = odf.size() / audioFileDuration;`

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirTypes.h:
```diff
@@ -0,0 +1,64 @@
+   double audioFileDuration = 0.;
+   std::vector<int> odfPeakIndices;
+   std::vector<float> odfAutoCorr;
+   std::vector<int> odfAutoCorrPeakIndices;
```
Comments at least for the more abbreviated member names?  Remind the reader what "odf" means

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
Good, I like to see the "IIFE" technique used to make clear that something remains const after once computed.

(Immediately Invoked Function Expression)

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,453 @@
+using PossibleDivHierarchies =
+   std::unordered_map<int /*num tatums*/, std::vector<BarDivision>>;
+
+// Gather a collection of possible number of divisions based on the assumption
```
"numbers"

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,453 @@
+                     const int numTatums = tatumsPerBar * numBars;
+                     const auto tatumRate = numTatums / audioFileDuration * 60;
+                     if (
+                        IsRound(tatumsPerBar) &&
```
I wonder about this use of `IsRound` sometimes wrongly rejecting a hypothesis.  Would it be better to do it with integer operations?

Suppose `tatumsPerBeat` contains nums and denoms,

`(numTatums * num) % denom == 0`

is the condition

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirTypes.h:
```diff
@@ -0,0 +1,64 @@
+{
+   OnsetQuantization tatumQuantization;
+   double bpm = 0;
+   std::optional<TimeSignature> timeSignature;
```
Maybe one `MusicalMeter` member

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
You assigned that already above

Maybe just collect all assignments here

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
Second argument should be called `numDivisions`

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirTypes.h:
```diff
@@ -0,0 +1,64 @@
+{
+   double error;
+   int lag;
+   int numDivisions;
```
I got confused reviewing elsewhere -- this member actually is a number of tatums, not divisions

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,453 @@
+   switch (barDivision.beatsPerBar)
+   {
+   case 2:
+      return (tatumsPerBeat == 3) ? TimeSignature::SixEight :
```
I feel a similar doubt here, as with `IsRound` -- you should express the test with integer arithmetic only

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
Do as in the previous suggestion, and `ts` will not be optional.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
So, this while loop is climbing up to a local peak?  Summarize with a comment if I understand right

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,159 @@
+      // Compress the frame as per section (6.5) in Müller, Meinard.
+      // Fundamentals of music processing: Audio, analysis, algorithms,
+      // applications. Vol. 5. Cham: Springer, 2015.
+      constexpr auto gamma = 100.f;
```
You don't need to capture this in the lambda -- it's a constant

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirDsp.cpp:
```diff
@@ -0,0 +1,159 @@
+      // larger this multiplier, the less peaks will remain. This value was
+      // found by trial and error, using the benchmarking framework
+      // (see TatumQuantizationFitBenchmarking.cpp)
+      constexpr auto thresholdRaiser = 1.5;
```
I've done bottom-up review of all these calculations, twice now.  I'm doing one more pass, top-down, depth first.

Now I think it would be clearer if moved this multiplication by 1.5, into MirDsp where the moving average is used.

Then this so-called `GetMovingAverage` will really do what the function name suggests, and only that.

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,453 @@
+constexpr auto maxBpm = 200.;
+constexpr auto minBeatsPerBar = 2;
+constexpr auto maxBeatsPerBar = 4;
+constexpr std::array<int, 3> possibleBeatsPerBar { 2, 3, 4 };
```
unused

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -0,0 +1,453 @@
+// Gather a collection of possible number of divisions based on the assumption
+// that the audio is a loop (hence there must be a round number of bars) and on
+// reasonable bar and tatum durations.
+PossibleDivHierarchies GetPossibleDivHierarchies(double audioFileDuration)
```
Maybe this function could be called `GenerateTempoHypotheses`

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
This is used only in `RunQuantizationExperiment` -- maybe this computation could be lowered into that function for clarity; pass `possibleDivs` into it instead

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
"Generate hypotheses" here about tempo...

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
Maybe call this "Best fit tatum number"

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
Could there be more than one hypthetical bar division with a four/four signature?  If so, why should the first be preferred?

saintmatthieu replied:
Absolutely, thanks.

saintmatthieu commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
The autocorrelation is not conjugate-symmetric, only symmetric: it's real numbers. I'm confident that the code below is correct.

saintmatthieu commented:
@Paul-Licameli I've addressed the 3 things we've agreed on: the `div / 8` error, filtering 4/4 time signatures rather than mistakenly just taking the first of them, and using integer arithmetic when filtering out hypotheses that don't have a round number of tatums per bar.
I've also addressed some of the easiest other comments. Sorry at this late hour I don't have sufficient brains for more.
And finally, I've rebased and put aside the benchmarking commits for now. Since these _do not_ change the behaviour of Audacity, I think they can be pushed even while QA is doing its work. But for today they're out.
I'll make time tomorrow to address more of your if I'm confident they don't change behaviour. My benchmarking tests can help me verifying this.

Paul-Licameli commented:
Reminding ourselves that https://github.com/audacity/audacity/pull/5791#discussion_r1441172648 points to a pre-existing problem that should have another issue and will be easily fixed.

Paul-Licameli commented on libraries/lib-music-information-retrieval/MirUtils.h:
```diff
@@ -26,9 +26,8 @@ std::vector<int> GetPeakIndices(const std::vector<float>& x);
- * dealing with symmetric spectra reduced only to their positive half.
- * See tests below for more details.
+ * @brief Useful when dealing with symmetric spectra reduced only to their
+ * positive half. See tests below for more details.
```
Incidentally noticed that some other functions in `MirUtils` are not used at all

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -347,7 +347,7 @@ MusicalMeter GetMostLikelyMeterFromQuantizationExperiment(
    std::vector<BarDivision> fourFourDivs;
    for_each_in_range(
-      IotaRange { 0u, possibleBarDivisions.size() }, [&](size_t i) {
+      IotaRange<size_t>(0, possibleBarDivisions.size()), [&](size_t i) {
```
C++23 would let you write `0uz` instead to make a size_t literal

Paul-Licameli commented on libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp:
```diff
@@ -342,23 +342,28 @@ size_t GetBestBarDivisionIndex(
-   // If there is a possibility that this is a 4/4, we don't take any risk and
-   // use this. When we get more clever and use the beat contents, we may be
+
+   // If there is one or more 4/4 possibilities, don't take any risk and only
```
are not is

Paul-Licameli reviewed (approved):
Good enough to move on to QA!  I understand there will be other later issues to fix the undo problem, improve quality and performance, and add unit tests.

I would prefer a bit of squashing before merge of course.

Paul-Licameli commented:
Does the last commit affect caching or the missed beats in time track?

saintmatthieu commented:
> Does the last commit affect caching or the missed beats in time track?

It's only about caching. What missed beats do you mean?

saintmatthieu commented:
@Paul-Licameli I cherry-picked the benchmarking test code onto that branch (including your suggestion commits). Production code is not modified at all so that won't interfere with QA. A lot of follow-up work depends on this and I already have a benchmarking-improvement ticket, so didn't want to add to the confusion by creating yet another similarly named ticket.

saintmatthieu commented on libraries/lib-music-information-retrieval/tests/MirTestUtils.h:
```diff
@@ -0,0 +1,162 @@
+
+// Some tests, such as benchmarking and visualization, are not meant to be run
+// on CI. This variable is used to disable them.
+static constexpr auto runLocally = false;
```
Comment from @Paul-Licameli:
> So I understand you change this to true, run locally, and then decide just two magic number thresholds to put into the application, in MusicInformationRetrieval.h. The very last thing in GetMeterUsingTatumQuantizationFit.cpp is use of those thresholds.
>
> Maybe that header file should have a comment about how the numbers are experimentally derived.

saintmatthieu replied:
ok

saintmatthieu commented on libraries/lib-music-information-retrieval/tests/MirTestUtils.h:
```diff
@@ -0,0 +1,162 @@
+      return { auc, results.back().score };
+   else if (it == falsePositiveRates.begin())
+      // No breakpoint satisfies the constraint. Return the greatest score.
+      return { auc, results.front().score };
```
From @Paul-Licameli:
> I think this branch should return greatest score plus some epsilon.

saintmatthieu replied:
It could, but since the false-positive tolerance itself being rather arbitrary, we shouldn't be preoccupied with that kind of precision.

saintmatthieu commented on libraries/lib-music-information-retrieval/tests/MirTestUtils.h:
```diff
@@ -0,0 +1,162 @@
+// Reproducible benchmarking must use the same input. We use this to make sure
+// that it does.
+template <int bufferSize = 1024>
+void UpdateHash(const MirAudioReader& source, float& hash)
```
From @Paul-Licameli:
> Maybe write it in functional style, taking a value and returning a new value.
>
>I lift eyebrows to see a floating point type for a "hash" value. I looked for uses of this function. You are not really using it in a standard unordered container, which would require some quantization of the value to size_t. I think "checksum" better describes the use of the value.

saintmatthieu replied:
I agree, checksum is a better choice of name.
Also changing the signature to something more straightforward.

saintmatthieu commented on libraries/lib-music-information-retrieval/tests/TatumQuantizationFitBenchmarkingOutput/summary.txt:
```diff
@@ -0,0 +1,5 @@
+Strict Threshold (Minutes-and-Seconds): 0.8367879995455243
+Lenient Threshold (Beats-and-Measures): 0.7336582045210642
+Octave error RMS: 0.1248861169665804
+Audio file hash: -1840.31005859375
```
From @Paul-Licameli:
> That's strange -- I thought the checksum would be a summation of samples of audio files, which would tend to average toward 0.

saintmatthieu replied:
Looks ok. I've just run the benchmarking again, counting the total number of samples, which is 84540711. To get the average DC offset in dB, `20*log10(1840.31005859375/84540711) == -93.24349792273436`, close to the noise floor for 16bit audio.

saintmatthieu commented on libraries/lib-music-information-retrieval/tests/TatumQuantizationFitBenchmarking.cpp:
```diff
@@ -0,0 +1,220 @@
+   // other developers is currently being worked on.
+
+   // We only observe the results for the most lenient classifier. The other,
+   // stricter classifier will yield the same results, only with less false
```
From @Paul-Licameli:
> fewer -- sorry to be pedantic

saintmatthieu replied:
I don't mind - thanks.

saintmatthieu commented on libraries/lib-music-information-retrieval/tests/MirTestUtils.cpp:
```diff
@@ -0,0 +1,83 @@
+   std::cout.flush();
+}
+
+OctaveError GetOctaveError(double expected, double actual)
```
From @Paul-Licameli:
https://github.com/audacity/audacity/pull/5843#discussion_r1443255418
(Comment is a bit long to be copied.)

saintmatthieu commented on libraries/lib-music-information-retrieval/tests/TatumQuantizationFitBenchmarking.cpp:
```diff
@@ -0,0 +1,220 @@
+   // If this changed, then some non-refactoring code change happened. If
+   // `rocInfo.areaUnderCurve > previousAuc`, then there's probably no argument
+   // about the change. On the contrary, though, the change is either an
+   // inadverted bug, and if it is deliberate, should be well justified.
```
From @Paul-Licameli:
> inadvertent

saintmatthieu commented on libraries/lib-music-information-retrieval/tests/TatumQuantizationFitBenchmarking.cpp:
```diff
@@ -0,0 +1,220 @@
+   // `rocInfo.areaUnderCurve > previousAuc`, then there's probably no argument
+   // about the change. On the contrary, though, the change is either an
+   // inadverted bug, and if it is deliberate, should be well justified.
+   REQUIRE(rocInfo.areaUnderCurve == previousAuc);
```
From @Paul-Licameli:
> I think you might not get a precise match of numbers if you run the tests an a different platform or CPU

saintmatthieu replied:
That could be a problem. I expect it will be close, though, since I don't expect the false-positive and true-positive rates to differ much or at all. As for the floating-point arithmetic ...
In a follow-up PR I intend to make the evaluation dataset easily available, not only on the machine I'm using right now. When I'm there I'll test it on the macbook I have, see what happens.

saintmatthieu commented on libraries/lib-music-information-retrieval/tests/TatumQuantizationFitBenchmarking.cpp:
```diff
@@ -0,0 +1,220 @@
+               << "AUC: " << rocInfo.areaUnderCurve << "\n"
+               << "Strict Threshold (Minutes-and-Seconds): " << strictThreshold
+               << "\n"
+               << "Lenient Threshold (Beats-and-Measures): "
```
From @Paul-Licameli:
> I notice the checked-in summary.txt was not generated by this version of the .cpp file

saintmatthieu replied:
Yes, the commits I'm preparing should fix this. Most importantly, the calculated thresholds are consistent with the values used in `MusicInformationRetrieval.h`,
```
static const std::unordered_map<FalsePositiveTolerance, LoopClassifierSettings>
   loopClassifierSettings {
      { FalsePositiveTolerance::Strict, { .01, 0.8367879995455243 } },
      { FalsePositiveTolerance::Lenient, { .1, 0.7336582045210642 } },
   };
   ```

saintmatthieu reviewed (commented):
@Paul-Licameli I copied over the comments you wrote on the fork you made of this branch.

dozzzzer commented:
@saintmatthieu 
When importing a non-looped rhythmical audio file, it still tries to detect its tempo even though it shouldn't.
So, to illustrate this misbehavior let's take [this loop of 78 bpm](https://www.dropbox.com/scl/fi/jarjq2jo22365jejvzbqp/bb_drum_break_loop_brooklyn1_78.wav?rlkey=nt9g6tug4kkczr2iywfxm4bzu&dl=0) as a source and trim it by about 1/16 from the right. The resulting loop would look smth[ like this](https://www.dropbox.com/scl/fi/pkp5l5i012x1viyw1f4f4/cut-off-loop.wav?rlkey=v59ce3mada9peqziy4jthxyhw&dl=0).
Import it into Audacity with "Ask me each time" in Preferences, and there you have it: Audacity falsely detects the tempo.
 
<img width="400" alt="image" src="https://github.com/audacity/audacity/assets/70063970/719e541c-89ea-4d83-8d4a-9e33c995f159">

saintmatthieu commented:
Yes: the algorithm detects it as 7 bars of 3/4. There probably is such a periodicity in the onset pattern, which is all the algorithm takes into account at the moment. If it also looked at timbre change patterns, it could probably realize that there in fact is no such periodicity. With a bit of luck I can work on that, but later.
